### PR TITLE
Define and enforce canonical ordering for DKG and protocol state

### PR DESCRIPTION
**Pull request by jordanschalm**
This PR defines and enforces a canonical ordering for identity lists when running the DKG and within the protocol state.

### CODE REVIEW

File: cmd/bootstrap/cmd/finalize.go [hunk 1]
```diff
@@ -383,7 +383,10 @@ func internalStakesByAddress() map[string]uint64 {
 }
 
 // mergeNodeInfos merges the inernal and partner nodes and checks if there is no
```
Comment (tarakby) [inline cmd/bootstrap/cmd/finalize.go:385]:
```suggestion
// mergeNodeInfos merges the internal and partner nodes and checks if there are no
```

File: model/flow/identity.go [hunk 2]
```diff
@@ -274,6 +274,18 @@ func (il IdentityList) Order(less IdentityOrder) IdentityList {
 	return dup
 }
 
+// Ordered returns whether the list is ordered by the input ordering.
+func (il IdentityList) Ordered(less IdentityOrder) bool {
```
Comment (tarakby) [inline model/flow/identity.go:278]: Is it assumed that all elements of the receiver `il` are valid elements of the ordering `less`? I mean is the call `less(a,b)` always safe?

**Review by tarakby — APPROVED** [review]
Looks good to me 👌

File: model/flow/identity.go (hunk 2 above)
Comment (jordanschalm) [reply model/flow/identity.go:278]: It is safe. Orderings should be at least consistent when applied to any identity list (they are just a function `(a, b Identity) -> Bool`)

File: model/bootstrap/node_info.go [hunk 3]
```diff
@@ -220,6 +221,15 @@ func FilterByRole(nodes []NodeInfo, role flow.Role) []NodeInfo {
 	return filtered
 }
 
+// Order sorts the NodeInfo list in place using the given ordering.
+func Order(nodes []NodeInfo, order flow.IdentityOrder) []NodeInfo {
```
Comment (yhassanzadeh13) [inline model/bootstrap/node_info.go:225]: Maybe refactoring `Order` to `Sort` would reduce the cognitive load of what does it mean:
```
func Sort(nodes []NodeInfo, sorter flow.IdentityOrder) []NodeInfo
```

File: state/protocol/badger/state_test.go [hunk 4]
```diff
@@ -173,63 +174,68 @@ func TestBootstrapNonRoot(t *testing.T) {
...
+	t.Run("non-canonical ordering", func(t *testing.T) {
+		participants := unittest.IdentityListFixture(20, unittest.WithAllRoles())
 
-	root := unittest.RootSnapshotFixture(participants)
-	bootstrap(t, root, func(state *bprotocol.State, err error) {
-		assert.Error(t, err)
+		root := unittest.RootSnapshotFixture(participants)
+		// randomly shuffle the identities so they are not canonically ordered
+		encodable := root.Encodable()
+		encodable.Identities = encodable.Identities.DeterministicShuffle(time.Now().UnixNano())
```
Comment (yhassanzadeh13) [inline state/protocol/badger/state_test.go:234]: wondering why we can't directly do the following? looks like it follows the same logic 🤔 

```
encodable.Identities = participants.DeterministicShuffle(time.Now().UnixNano())
```

File: state/protocol/badger/state_test.go [hunk 5]
```diff
@@ -173,63 +174,68 @@ func TestBootstrapNonRoot(t *testing.T) {
...
+				participants := unittest.IdentityListFixture(5, unittest.WithAllRolesExcept(role))
+				root := unittest.RootSnapshotFixture(participants)
+				bootstrap(t, root, func(state *bprotocol.State, err error) {
+					assert.Error(t, err)
+				})
 			})
+		}
+	})
+
+	t.Run("existing address", func(t *testing.T) {
```
Comment (yhassanzadeh13) [inline state/protocol/badger/state_test.go:217]:
```suggestion
	t.Run("duplicate address", func(t *testing.T) {
```

**Review by yhassanzadeh13 — APPROVED** [review]

File: model/bootstrap/node_info.go (hunk 3 above)
Comment (jordanschalm) [reply model/bootstrap/node_info.go]: 👍  I agree that is clearer. Updated this method as well as the `IdentityList` methods in https://github.com/onflow/flow-go/pull/678/commits/2d834ed5d3da080cef5171006afd40045f5262cd

**Pull request merged**